PIPELINE: Fix config bug

nielsboecker opened the pull request:
Boolean and numerical values were parsed as strings, d'oh

nielsboecker committed:
PIPELINE: Fix config parsing bug

Boolean and numerical values were parsed as strings, d'oh

nbckr commented:
Note: There are some other places in the code where I read the numeric variables and cast them, which is now redundant. I will fix this in the othher branch I'm working on atm. I excluded it here as I made some other changes in these files and wanted to keep this PR here small.

nielsboecker committed:
PIPELINE: Remove redundant value casts of config variables

nielsboecker committed:
PIPELINE: Replace in-memory global state with file-based solution

Keeping state in-memory is hard with concurrency. At the moment we
have multiple worker processes in job_controller, and potentially
another level of concurrenty when gunicorn runs Flask. Also, this
is more in line with the rest of the state management.

nielsboecker committed:
PIPELINE: Update README.md to reflect new state-keeping method

nielsboecker committed:
PIPELINE: Add support for reading logs from finished jobs

boonwj commented on HoloPipelines/jobs/jobs_state.py:
```diff
@@ -34,32 +41,33 @@
+def update_job_state(job_id: str, new_state: str, logger=logging, new: bool = False):
     """
-    Updates the global dictionary that keeps track of all jobs. Note that new_state
+    Updates the state for aa job by updating the job.state file. Note that new_state
```
typo "a"

boonwj commented on HoloPipelines/jobs/jobs_io.py:
```diff
@@ -10,55 +10,106 @@
+    Reads state file and returns state and time in seconds since last modification.
+    """
+    state_file_path = get_state_file_path_for_job(jod_id)
+    with open(state_file_path, "r") as state_file:
```
There needs to be error handling here, as querying a non-existent job id will throw an exception

nielsboecker replied:
Agree, did it in the other branch

boonwj commented on HoloPipelines/jobs/jobs_io.py:
```diff
@@ -80,16 +131,28 @@ def get_logger_for_job(job_id: str):
+        return f"{path}/job.log"
+    elif os.path.isdir(path.replace(jobs_root, finished_jobs_root)):
+        return f"{path.replace(jobs_root, finished_jobs_root)}/job.log"
+
```
An explicit `return None` will be idiomatic given that the function returns different values and None is one of them

boonwj commented on HoloPipelines/jobs/jobs_io.py:
```diff
@@ -10,55 +10,106 @@
+        state_file.write(state)
 
 
 def get_logger_for_job(job_id: str):
```
Optional: type hints for return value?

boonwj commented on HoloPipelines/jobs/jobs_state.py:
```diff
@@ -5,17 +5,24 @@
+    state_file_for_job_exists,
+)
 
 JobState = Enum(
```
`JobState` does not feel like a proper python variable naming convention for global variables. The naming convention should be similar to python functions.

https://www.python.org/dev/peps/pep-0008/#global-variable-names

nielsboecker replied:
Hm but it is an Enum right? See for instance how they use it in the docs: https://docs.python.org/3/library/enum.html#functional-api

boonwj reviewed (approved):
No reason to reject the config bug fixes as this is part of the major refactoring changes. 

In addition, this PR also introduces changes to the way job states are tracked and managed.

It's best to merge it with the major-refactor branch and be evaluated as a whole.
